src/openrct2/peep/Peep.cpp

```diff
@@ -3068,22 +3068,60 @@ void Peep::PerformNextAction(uint8_t& pathing_result, TileElement*& tile_result)
 
     auto newLoc = *loc;
     CoordsXY truncatedNewLoc = newLoc.ToTileStart();
+    int16_t height = GetZOnSlope(newLoc.x, newLoc.y);
     if (truncatedNewLoc == CoordsXY{ NextLoc })
     {
-        int16_t height = GetZOnSlope(newLoc.x, newLoc.y);
         MoveTo(newLoc.x, newLoc.y, height);
         return;
     }
 
-    if (newLoc.x < 32 || newLoc.y < 32 || newLoc.x >= gMapSizeUnits || newLoc.y >= gMapSizeUnits)
+    //check if peep is on peep spawn location
+    bool outside = false;
+    if (newLoc.x <= 32 || newLoc.y <= 32 || newLoc.x >= gMapSizeUnits || newLoc.y >= gMapSizeUnits)
+        // apply same old behaviour
+        outside = true;
+
+    else
     {
-        if (outside_of_park == 1)
+        for (auto& elem : gPeepSpawns)
         {
-            pathing_result |= PATHING_OUTSIDE_PARK;
+            CoordsXYZ coord1, coord2;
+            coord1.x = elem.ToTileCentre().x;
+            coord1.y = elem.ToTileCentre().y;
+            coord1.z = elem.ToTileCentre().z;
+
+            CoordsXYZ tileCoord;
+            tileCoord.x = coord1.x / COORDS_XY_STEP;
+            tileCoord.y = coord1.y / COORDS_XY_STEP;
+            tileCoord.z = coord1.z;
+
+            // check if either peep is on the edge of the map or anywhere else
+            if (tileCoord.x == 0 || tileCoord.x == (gMapSizeUnits / COORDS_XY_STEP)-1 || tileCoord.y == 0
+                || tileCoord.y == (gMapSizeUnits / COORDS_XY_STEP)-1)
+            {
+                //do nothing
+            }
+
+            else if (
+                newLoc.ToTileCentre().x == coord1.x && newLoc.ToTileCentre().y == coord1.y && height == coord1.z
+                && direction == elem.direction)
```

Comment:
(not sure if it will be the final formatting, just simplifying the comparison)
```suggestion
            else if (newLoc.ToTileCentre() == coord1 && height == coord1.z && direction == elem.direction)
```